### ISSUE TITLE
rtl8723be: Weak signal on 4.12.x kernels

### ISSUE THREAD

**Issue by GDsouza** [lwfinger/rtlwifi_new#263]
Laptop: `HP Pavilion 15 - ab035ax`
Wireless Card: `RTL8723BE`
Options: `ant_sel=1`
Kernel: `4.12.3-1-ARCH`

The following three results are from the in-tree module

Touching the hotspot:
```
[root@EOS ~]# iw dev wlo1 scan | egrep "SSID|signal"
	signal: -66.00 dBm
	SSID: Glen
```

Approximately 1m away:
```
[root@EOS ~]# iw dev wlo1 scan | egrep "SSID|signal"
	signal: -74.00 dBm
	SSID: Glen
```

Approximately 2.5m away:
```
[root@EOS ~]# iw dev wlo1 scan | egrep "SSID|signal"
	signal: -86.00 dBm
	SSID: Glen
```

The signal rates given above remain consistent, for ant_sel=0 and 2, as well.

At this point, I decided to build the module from this git repository. The results of a quick test are tabulated below.

Testing with different ant_sel parameters for a fixed distance:

| ant_sel  | Signal  (dBm)  | Comments  |
|:---:|:---:| ---|
| 1      | -80.00      | Have been using this since kernel 4.7 |
| 0      | -80.00      | Has never given me good signal |
| 2      | -26.00      | Was used on kernel 4.6 and older |

The module compiled into the kernel does not work well (worse than -70 dBm), compiling the module from git yields proper functionality on `ant_sel=2`

I've had this issue before, where ant_sel values differ with kernel versions https://github.com/lwfinger/rtlwifi_new/issues/125

**Comment by bartbkr** [comment]
Same here after upgrading to kernel 4.12.3

**Comment by lwfinger** [comment]
We have reworked the btcoex code quite a bit, and I have no idea if the necessary value for ant_sel has stayed the same. It was introduced as a kludge to benefit those users that were screwed by their vendor.

Your signal values are too small. I will make some tests on the current kernel and report back.

**Comment by lwfinger** [comment]
OK, I have now done the tests that I had hoped one of you would do. I have full access to my device, which means I can change the antenna connection to be what I want. I tested rtlwifi_new, kernel mainline, and wireless-drivers-new. The distance to the AP is not relevant, just keep it constant. Running the sudo iw dev wlan0 scan | egrep "SSID|signal" command for the various combinations and kernels, I get the following:

rtlwifi_new           Antenna connected to connector #
ant_sel value                 1                            2 
0                                -38                         -66
1                                -31                         -66
2                                -78                         -34

mainline               Antenna connected to connector # 
ant_sel value                 1                             2
0                                 -38                          -64
1                                 -34                          -66
2                                 -34                          -86

wireless-drivers-next   Antenna connected to connector # 
ant_sel value                 1                             2
0                                 -34                          -58
1                                 -34                          -64
2                                 -42                          -62

Obviously, something has been broken, and the ant_sel no longer works for any other repo than rtlwifi_new. If that is what you were trying to tell me above, you certainly did not make it clear, and I have read so much crap in these issue comments that I barely read them any more.

The bottom line is that if you want me to read these reports, please make it clear what you are saying and plan your experiment carefully. If you want to show that there is a problem with the signal vs distance, then reporting the strength as a function of distance would be appropriate. If you want to state that the ant_sel code is not working, then doing that experiment merely shows me that you have no idea what you are doing!

**Comment by manjalyc** [comment]
I don't really understand how any of this works, but I went through and compiled and tested 23 kernels, each kernel was hard reset to one of the latest 23 commits in https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/log/drivers/net/wireless/realtek/rtlwifi/rtl8723be . The last commit that resulted in a working kernel (boots properly, strong signal, and connects to the internet) was [74a7dfbcc64986504dbf9f5f7c7cca521db4d230](https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/drivers/net/wireless/realtek/rtlwifi/rtl8723be?id=74a7dfbcc64986504dbf9f5f7c7cca521db4d230) from March 20th, 2017.

Here are the results, and some other information: https://docs.google.com/spreadsheets/d/14ZOXNOQR7gqibQlTeTEqzm8Hd6pSODIBl4vzTLPstqo/edit#gid=0

**Comment by lwfinger** [comment]
The next time you want to do this kind of debugging, you should investigate "git bisect". I found the bad commit with only 7 rebuilds. Not only that, but your limitation to patches to rtl8723be led you to the wrong commit. It is actually bcd37f4, which patched btcoexist. At the moment, the engineers and I are trying to determine where the error occurs. We should have it within a few days. It will take quite a while for the fix to propagate through the system and be applied to kernels 4.12 and 4.13. I suggest that you monitor the mailing list linux-wireless@vger.kernel.org to find the patch.

**Comment by manjalyc** [comment]
Oh ok, I knew there had to be a better way to go about this. I kind of just did this experiment to use git for the first time, but I'll be sure to use git bisect in the future.

**Comment by MrOutput** [comment]
Patch provided in #259.

**Comment by arunsai123** [comment]
Still having low signal on rtl8723be when using 4.14 kernel.

**Comment by GDsouza** [comment]
@arunsai123 Check your ant_sel parameter. This issue has been fixed

**Comment by arunsai123** [comment]
@GDsouza  I tried both ant_sel=1 and ant_sel=2. it didn't work. I am getting around only 42% when placed behind the router.
